lib/Corres_UL.thy

```diff
@@ -947,6 +947,11 @@ lemma corres_assert_assume_l:
   \<Longrightarrow> corres_underlying sr nf nf' rrel (P and (\<lambda>s. P')) Q (assert P' >>= f) g"
   by (force simp: corres_underlying_def assert_def return_def bind_def fail_def)
 
+lemma corres_assert_assume_l_forward:
+  "\<lbrakk>\<And>s. P s \<Longrightarrow> P'; corres_underlying sr nf nf' rrel (P and K P') Q (f ()) g\<rbrakk>
+  \<Longrightarrow> corres_underlying sr nf nf' rrel P Q (assert P' >>= f) g"
+  by (force simp: corres_underlying_def assert_def return_def bind_def fail_def)
```

Comment:
I'm puzzled by this. It appears if you use the lemma, you immediately get that `P s ==> P'` as your first goal, which means you'd have to specify P', or defer/rotate it and work on the corres_underlying first. What am I missing?

Reply:
This is a forwards style rule; I put `forward` in the name to indicate this, though as Gerwin pointed out recently, we do have many forwards style rules but not a proper naming convention.

So this rule would be used when the guards are concrete, not schematic. It will generate the first subgoal for you, which you immediately solve, and then you are left with a new `corres_underlying` goal, with concrete guards, and now with the property that was just asserted as a new conjunction in the Haskell guard.

Reply:
Yes, that's what I was missing. Thank you! All good.